File: c-sharp/Drawing/src/Color.cs
namespace RayTracer.Drawing
{
    public class Color
    {
        public float red { get; set; }
        public float green { get; set; }
        public float blue { get; set; }

        public Color(float red, float green, float blue)
        {
            this.red = red;
            this.green = green;
            this.blue = blue;
        }

        public Color(double red, double green, double blue)
            : this((float)red, (float)green, (float)blue)
        { }

        public static Color operator +(Color a, Color b)
        {
            return new Color(
                a.red + b.red,
                a.green + b.green,
                a.blue + b.blue
            );
        }

        public static Color operator -(Color a, Color b)
        {
            return new Color(
                a.red - b.red,
                a.green - b.green,
                a.blue - b.blue
            );
        }

        public static Color operator *(Color a, int i)
        {
            return new Color(
                a.red * i,
                a.green * i,
                a.blue * i
            );
        }

        public static Color operator *(Color a, Color b)
        {
            return new Color(
                a.red * b.red,
                a.green * b.green,
                a.blue * b.blue
            );
        }
    }
}

File: cpp/the-ray-tracer-challenge/src/Color.cpp
#include "Color.h"

Color::Color() : e{ 0, 0, 0 } {};

Color::Color(float r, float g, float b) : e{ r, g, b } {}

float Color::r() const { return e[0]; }

float Color::g() const { return e[1]; }

float Color::b() const { return e[2]; }

bool Color::operator==(const Color& rhs) const {
	return floatEqual(r(), rhs.r())
		&& floatEqual(g(), rhs.g())
		&& floatEqual(b(), rhs.b());
}

Color Color::operator+(const Color& rhs) const {
	return Color{ r() + rhs.r(), g() + rhs.g(), b() + rhs.b() };
}

Color Color::operator-(const Color& rhs) const {
	return Color{ r() - rhs.r(), g() - rhs.g(), b() - rhs.b() };
}

Color Color::operator*(const Color& rhs) const {
	return Color{ r() * rhs.r(), g() * rhs.g(), b() * rhs.b() };
}

Color Color::operator*(float scalar) const {
	return Color{ r() * scalar, g() * scalar, b() * scalar };
}